### ISSUE TITLE
Don't serve expired results from twisted.names.cache.CacheResolver

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5647]
|<img alt="acidfoo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| acidfoo reported|
|-|-|
|Trac ID|trac#5647|
|Type|enhancement|
|Created|2012-04-16 13:55:00Z|
|Branch|https://github.com/twisted/twisted/tree/ignore-expired-records-5647-2|

Instead of ensuring TTL's are > 0, CacheResolver should rather not serve results that are still in the cache even though they have expired. See previous discussion in [#5579](https://github.com/twisted/twisted/issues/5579)

Attachments:

* [expired.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cdc/cdc1d0df5ecef0be02ed3b14e936eadbc77ecc1e/2c5aa409bc0f45c2a1b130ed2813f7550b034999.diff) (5053 bytes) - added by acidfoo on 2012-04-16 13:55:19Z - 
* [therve-review-patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cdc/cdc1d0df5ecef0be02ed3b14e936eadbc77ecc1e/1c53c5eb323f29f5cd3deba08ecadd5a358accbb.diff) (4214 bytes) - added by acidfoo on 2012-06-02 10:37:11Z - 
* [merged-patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cdc/cdc1d0df5ecef0be02ed3b14e936eadbc77ecc1e/8ee8e49d3c2c05ee7a7d5fc3b896da03424b384b.diff) (189135 bytes) - added by acidfoo on 2012-06-02 10:37:27Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5647 5647
type__enhancement enhancement
reporter__acidfoo acidfoo
priority__normal normal
milestone__ 
branch__branches_ignore_expired_records_5647_2 branches/ignore-expired-records-5647-2
branch_author__acidfoo__exarkun acidfoo, exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1334584500000000 1334584500000000
changetime__1340747642000000 1340747642000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="acidfoo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|acidfoo commented|
|-|-|

The cleanest way of doing this ended up being to raise a ValueError when trying to create a RRHeader with a negative TTL. This is new behaviour in RRHeader but I don't think it would break any code that wasn't incorrect in the first place.

The other options where to perform another loop through all the records, or to make some sort of buildHeaderOrFail that wrapped the creation of RRHeader and checking of TTL. Neither of these ended up looking great, but I have the code if it's decided that RRHeader shouldn't change.

**Comment by twisted-trac** [comment]
|<img alt="acidfoo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|acidfoo commented|
|-|-|

Oh, this would also invalidate [#5636](https://github.com/twisted/twisted/issues/5636) since the unit test no longer expects a result back when the record has expired.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [34253]) Apply ~/Downloads/expired.diff 

refs [#5647](https://github.com/twisted/twisted/issues/5647)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [34252]) Branching to 'ignore-expired-records-5647'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to acidfoo|
|-|-|

1.There is a simple conflict in test_cache.py.
 
 2. There are some trailing whitespaces in test_cache.py and dns.py.

 3. The branch is missing a NEWS fragment.

 4. Please put full stops at the end of the different docstrings.

 5 Just for clarity, I would move the succeed out of the try/except, something like this:

```
try:
    result = [...]
except ValueError:
    return defer.fail(...)
else:
    return defer.succeed(result)
```

If you can provide a patch against the branch that'd be awesome. Thanks!

**Comment by twisted-trac** [comment]
|<img alt="acidfoo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|acidfoo commented|
|-|-|

Thanks for the review. I've made all the changes you suggested. I'm not sure
how to handle resolving the conflict so I'm going to attach two diffs - one that
is just the changes and the second that is a diff after I merged and resolved
the conflict. This was a merge against trunk revision 34477 
(svn merge -r 34253:34477 svn://svn.twistedmatrix.com/svn/Twisted/trunk)

In the actual conflict, we want to keep the code from this patch which is:
 	99	        return self.assertFailure( 
 	100	            c.lookupAddress("example.com"), dns.DomainError)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [34531]) Branching to 'ignore-expired-records-5647-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [34641]) Get rid of "should" from a test method docstring and minorly adjust some whitespace in another test.

refs [#5647](https://github.com/twisted/twisted/issues/5647)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

Thanks.  Finally got around to a second look at this.  I've just understood some potential TTL issues in Twisted Names, but they're not new in this branch, so they don't matter for now. :)   Apart from the two minor issues fixed in the revision linked above, this looks good to me now.  I'll go ahead and merge it.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [34642]) Merge ignore-expired-records-5647-2

Author: acidfoo
Reviewer: therve, exarkun
Fixes: [#5647](https://github.com/twisted/twisted/issues/5647)

Completely avoid serving expired records from the Twisted Names cache
resolver, rather than pegging the TTL of expired records to 0 and serving
them anyway.
